src/Data/Array/NonEmpty.purs

```diff
@@ -0,0 +1,327 @@
+module Data.Array.NonEmpty where
+
+import Prelude
+
+import Control.Alt (class Alt)
+import Control.Alternative (class Alternative)
+import Control.Lazy (class Lazy)
+import Control.Monad.Rec.Class (class MonadRec)
+import Data.Array as A
+import Data.Bifunctor (bimap)
+import Data.Eq (class Eq1)
+import Data.Foldable (class Foldable)
+import Data.FoldableWithIndex (class FoldableWithIndex)
+import Data.FunctorWithIndex (class FunctorWithIndex)
+import Data.Maybe (Maybe(..), fromJust)
+import Data.NonEmpty (NonEmpty, (:|))
+import Data.Ord (class Ord1)
+import Data.Traversable (class Traversable)
+import Data.TraversableWithIndex (class TraversableWithIndex)
+import Data.Tuple (Tuple)
+import Data.Unfoldable (class Unfoldable)
+import Partial.Unsafe (unsafePartial)
+
+newtype NonEmptyArray a = NonEmptyArray (Array a)
+
+instance showNonEmptyArray :: Show a => Show (NonEmptyArray a) where
+  show (NonEmptyArray xs) = "(NonEmptyArray " <> show xs <> ")"
+
+derive newtype instance eqNonEmptyArray :: Eq a => Eq (NonEmptyArray a)
+derive newtype instance eq1NonEmptyArray :: Eq1 NonEmptyArray
+
+derive newtype instance ordNonEmptyArray :: Ord a => Ord (NonEmptyArray a)
+derive newtype instance ord1NonEmptyArray :: Ord1 NonEmptyArray 
+
+
+derive newtype instance functorNonEmptyArray :: Functor NonEmptyArray
+derive newtype instance functorWithIndexNonEmptyArray :: FunctorWithIndex Int NonEmptyArray
+
+derive newtype instance foldableNonEmptyArray :: Foldable NonEmptyArray
+derive newtype instance foldableWithIndexNonEmptyArray :: FoldableWithIndex Int NonEmptyArray
+
+derive newtype instance traversableNonEmptyArray :: Traversable NonEmptyArray
+derive newtype instance traversableWithIndexNonEmptyArray :: TraversableWithIndex Int NonEmptyArray
+
+derive newtype instance applyNonEmptyArray :: Apply NonEmptyArray
+
+derive newtype instance applicativeNonEmptyArray :: Applicative NonEmptyArray
+
+derive newtype instance bindNonEmptyArray :: Bind NonEmptyArray
+
+derive newtype instance monadNonEmptyArray :: Monad NonEmptyArray
+
+derive newtype instance altNonEmptyArray :: Alt NonEmptyArray
+
+-- | Internal - adapt an Array transform to NonEmptyArray
+adapt :: forall a b. (Array a -> Array b) -> NonEmptyArray a -> NonEmptyArray b
```

Comment:
I realise this is internal, but even so I'm a little uneasy about defining this function with this type, as it's actually partial; if the function you provide can return an empty array you're in trouble. I think it might be better to use `unsafeCoerce` in individual implementations within this module; it's then easier (imo) to verify what's going on and that the coercion is justifiable.

Reply:
Good point. Thanks.

Reply:
For the moment, I'm going to make a small concession on this and just rename these to `unsafeAdapt`. I can always go back and do the `unsafeCoerce` option later. That said, I do think that `unsafeAdapt`, along with docs, should tell anyone hacking on the library what to look for / be aware of.